### DISCUSSION TITLE
Playing DSD on iOS

### DISCUSSION THREAD

**Discussion by doriansgithub** [sbooth/SFBAudioEngine#264] [Q&A]
Hi there, I need a clue to get this thru.

The sample uses a local file with NSString *filePath = [[NSBundle mainBundle] pathForResource:@"test" ofType:@"Flac"], which I am able to play.  

But I can't figure out how to get it to play from an array with different file types including dsd, dsf, etc. 

My code filters the array and feeds it into this method, at which point the player simply has to play the url.

- (void) setLocalSongPlay: (NSArray *) localArray
				mediaItem: (MPMediaItem *) item {
    
	self.arrSongs = localArray.mutableCopy;
	self.urlToPlay = item.assetURL;

I've tried this but it won't play:
        [sfbAudioPlayer playURL:self.urlToPlay error:nil];

I've tried this but it crashes:
        [sfbAudioPlayer playDecoder: (SFBDSDPCMDecoder*) self.urlToPlay error:nil];

Any ideas or sample?

**Comment by sbooth** [comment]
What do you expect `[sfbAudioPlayer playDecoder: (SFBDSDPCMDecoder*) self.urlToPlay error:nil];` to do? What is the anticipated result of casting to an unrelated type?

Please read [How do I ask a good question](https://stackoverflow.com/help/how-to-ask) and also the documentation (in the headers) for SFBAudioEngine. I believe that the answers you need are there, but without sufficient detail on what you're attempting and a reproducible example it's difficult to know the root cause. Some of the issues you're having seem to be unrelated to SFBAudioEngine at first glance.

I also recommend never passing `nil` for errors because they will often illuminate the cause of problems.

**Comment by doriansgithub** [comment]
It must be related to SFR as it plays without issues with the stk player. But the Sri player can’t play dsd files. This is why I want to use fsb.Neither my friend programmer or myself seem to get fsb to play an url without explicitly telling it the file name and extension.  And even then, if it’s a dsd file, it won’t play it.It plays the flac file, but again, I have to tell it the name and extension for it to play. That’s how it is in the sample.Ok, I’ll change the error.Sent from my iPhoneOn Jun 19, 2023, at 6:03 AM, Stephen Booth ***@***.***> wrote:﻿
What do you expect [sfbAudioPlayer playDecoder: (SFBDSDPCMDecoder*) self.urlToPlay error:nil]; to do? What is the anticipated result of casting to an unrelated type?
Please read How do I ask a good question and also the documentation (in the headers) for SFBAudioEngine. I believe that the answers you need are there, but without sufficient detail on what you're attempting and a reproducible example it's difficult to know the root cause. Some of the issues you're having seem to be unrelated to SFBAudioEngine at first glance.
I also recommend never passing nil for errors because they will often illuminate the cause of problems.

**Comment by doriansgithub** [comment]
If I am trying to play a dsd file, it seemed like a good idea to use the dsd decoder.Sent from my iPhoneOn Jun 19, 2023, at 9:21 AM, Dorian Mattar ***@***.***> wrote:﻿It must be related to SFR as it plays without issues with the stk player. But the Sri player can’t play dsd files. This is why I want to use fsb.Neither my friend programmer or myself seem to get fsb to play an url without explicitly telling it the file name and extension.  And even then, if it’s a dsd file, it won’t play it.It plays the flac file, but again, I have to tell it the name and extension for it to play. That’s how it is in the sample.Ok, I’ll change the error.Sent from my iPhoneOn Jun 19, 2023, at 6:03 AM, Stephen Booth ***@***.***> wrote:﻿
What do you expect [sfbAudioPlayer playDecoder: (SFBDSDPCMDecoder*) self.urlToPlay error:nil]; to do? What is the anticipated result of casting to an unrelated type?
Please read How do I ask a good question and also the documentation (in the headers) for SFBAudioEngine. I believe that the answers you need are there, but without sufficient detail on what you're attempting and a reproducible example it's difficult to know the root cause. Some of the issues you're having seem to be unrelated to SFBAudioEngine at first glance.
I also recommend never passing nil for errors because they will often illuminate the cause of problems.

**Comment by sbooth** [comment]
In addition to the error message some information that will help narrow things down:
- What is the source of the audio you're trying to play?
- Is the URL you're trying to play accessible using `fopen` or similar from within the app?
- What format is the audio?
- What device are you using for output? 
- What formats does it support?
- Do you want to play DSD using DoP or converted to PCM?

**Comment by doriansgithub** [comment]
The source is audio in the iPhone. The music app as well as my app using stk player are able to play them.I don’t use fopen, I have a media picker.Different types, including AIFF, dsd, mp3, m4a, Ape, flac. First I just want to be able to play the tracks. Then be able to play the best quality possible depending on the output; usb, Bluetooth, airplay, etc.The player that I currently use supports everything except dsd.I’m not sure what is best, I will have to resécate what is better. Do you have a suggestion between DoP or PCM?I just want the best possible audio in the app.Thanks.Sent from my iPhoneOn Jun 19, 2023, at 10:36 AM, Stephen Booth ***@***.***> wrote:﻿
In addition to the error message some information that will help narrow things down:

What is the source of the audio you're trying to play?
Is the URL you're trying to play accessible using fopen or similar from within the app?
What format is the audio?
What device are you using for output?
What formats does it support?
Do you want to play DSD using DoP or converted to PCM?

**Comment by doriansgithub** [comment]
For the test, I added a did file to the project and I’m feeding it directly 

NSString *filePath = [[NSBundle mainBundle] pathForResource:@"01 - Back In The Saddle” ofType:@"dsf”];

**Comment by doriansgithub** [comment]
Ok, got this error.

SFB Error: The type of the file “01 - Back In The Saddle.dsf” is not supported.

So what do I need to do to have it play .dsf files?

Thanks!

**Reply by sbooth** [reply]
What is the associated log message? That will have specific information on the chunk causing the issue.

**Comment by doriansgithub** [comment]
On the Mac sample, this is used before playing:

	/// Returns a decoder for this playlist item or `nil` if the audio type is unknown
	func decoder(enableDoP: Bool = false) throws -> PCMDecoding? {
		let pathExtension = url.pathExtension.lowercased()
		if AudioDecoder.handlesPaths(withExtension: pathExtension) {
			return try AudioDecoder(url: url)
		} else if DSDDecoder.handlesPaths(withExtension: pathExtension) {
			let dsdDecoder = try DSDDecoder(url: url)
			return enableDoP ? try DoPDecoder(decoder: dsdDecoder) : try DSDPCMDecoder(decoder: dsdDecoder)
		}
		return nil
	}

So do I need to run this before playing on iOS?

Thanks,

Dorian

**Reply by sbooth** [reply]
It depends on your device and its capabilities and what you're trying to accomplish. I'm guessing you want to convert the DSD to PCM so use a `DSDPCMDecoder`.

**Comment by doriansgithub** [comment]
The DSDPCMDecoder is in the iOS sample, but it’s not actually implemented.

I have tried making it work, but I can’t.  My friend is also trying and will let me know tomorrow if he can get it to work.

But just wondering why they have the code to decode, but don’t actually use it.

It works great in the Mac sample.

But the iOS returns a not supported format for DSD files.

**Reply by sbooth** [reply]
`DSDPCMDecoder` is definitely available in iOS. SimplePlayer doesn't use it but the intent of the demo is to illustrate use of the framework rather than be a full-fledged player.

**Comment by doriansgithub** [comment]
I understand that, not asking for a full-fledged player, but a sample on how to use the decoder would save a lot of time.

Anyways, thanks.

**Reply by sbooth** [reply]
I just pushed https://github.com/sbooth/SFBAudioEngine/commit/14b98653b2688fa27206e6888bee494abaa772a6 which should enable DSD-to-PCM decoding where possible for the iOS version of SImplePlayer.

**Comment by doriansgithub** [comment]
Y D Man!Will try it.Thanks!Sent from my iPhoneOn Jun 21, 2023, at 8:47 AM, Stephen Booth ***@***.***> wrote:﻿
I just pushed 14b9865 which should enable DSD-to-PCM decoding where possible for the iOS version of SImplePlayer.

**Comment by doriansgithub** [comment]
Hi, I am only able to play non dsd/dsf files.

```
// TESTING
        NSString *filePath = [[NSBundle mainBundle] pathForResource:@"01 - Back In The Saddle"
                                                             ofType:@"dsf"];
        
        self.urlToPlay = [NSURL fileURLWithPath:filePath];
        NSString *pathExtension = [[_urlToPlay pathExtension] lowercaseString];

        [sfbAudioPlayer playURL:_urlToPlay
                          error: &error];
        
        if ([SFBAudioDecoder handlesPathsWithExtension:pathExtension]) {

            NSLog(@"decoder Player %@", _urlToPlay);

                        [sfbAudioPlayer playURL:_urlToPlay
                                          error: &error];

        } else if ([SFBDSDDecoder handlesPathsWithExtension:pathExtension]) {

            SFBDSDDecoder *dsdDecoder = [[SFBDSDDecoder alloc] initWithURL:_urlToPlay
                                                                     error:nil];
            [[SFBDSDPCMDecoder alloc] initWithDecoder:dsdDecoder
                                                error:nil];
            
                        [sfbAudioPlayer playURL:_urlToPlay
                                          error: &error];
        }
```

**Reply by sbooth** [reply]
It looks like you're calling `playURL:error:` twice, once before the conditional block and once within. I'd start by removing the play call outside the block. What is the error message and what is in the log? That information is key.

**Comment by doriansgithub** [comment]
This screen capture shows that there is an mp3 track in the URL, but the player won’t play it.  It only play the test.flac file that is embedded in the app.

￼

**Reply by sbooth** [reply]
@doriansgithub you haven't provided adequate information to diagnose the cause. "won't play it" is not meaningful from a debugging perspective. What is the error message? What log messages are shown?

**Comment by doriansgithub** [comment]
I’m not getting errors.

￼

**Comment by doriansgithub** [comment]
I got this now.

error	NSError *	domain: @"org.sbooth.AudioEngine.AudioDecoder" - code: 2	0x0000000283c95140

**Comment by doriansgithub** [comment]
Hi, I am able to play dsd files now, but my local ipodlibrary files will not play unless I feed the file manually.

This is the code that makes it work.

Now we are working on playing straight from the _urlToPlay.


        NSString *pathExtension = [[_urlToPlay pathExtension] lowercaseString];
        
        if ([SFBAudioDecoder handlesPathsWithExtension:pathExtension]) {
            //mp3, ma4, aif, flac
            
            NSLog(@"decoder Player %@", _urlToPlay);
            [sfbAudioPlayer playURL:_urlToPlay
                              error: &error];
            NSLog(@"sfbAudioPlayer Error %@", error);

        } else if ([SFBDSDDecoder handlesPathsWithExtension:pathExtension]) {
            //dsd, dsf, flac
            SFBDSDDecoder *dsdDecoder = [[SFBDSDDecoder alloc] initWithURL:_urlToPlay
                                                                     error:&error];
            enableDoP = false;
            SFBDoPDecoder *objURLdop;
            SFBDSDPCMDecoder *objURLpcm;
                    if (enableDoP) {
                        NSLog(@"decoderDoP %@", _urlToPlay);
                        objURLdop = [[SFBDoPDecoder alloc] initWithDecoder:dsdDecoder
                                                            error:&error];
                        NSLog(@"SFBDSDdopDecoder Error %@", error);

                    } else {
                        NSLog(@"decoderPCM %@", _urlToPlay);
                        objURLpcm = [[SFBDSDPCMDecoder alloc] initWithDecoder:dsdDecoder
                                                            error:&error];
                        NSLog(@"SFBDSDpcmDecoder Error %@",error);
                    }

            [sfbAudioPlayer playDecoder:objURLpcm error:&error];
            NSLog(@"sfbAudioPlayer enqueueURL Player %@", error);
        }
